OSDOCS-20040-2: Adds PlanPolicy to RHCL (#114514)
openshift/openshift-docs: modules/proc-rhcl-configuring-authentication-planpolicy.adoc

// Module included in the following assemblies:
//
// *deployment/rhcl-planpolicy.adoc

:_mod-docs-content-type: PROCEDURE
[id="proc-rhcl-configuring-authentication-planpolicy_{context}"]
= Creating an AuthPolicy with API key authentication

[role="_abstract"]

> You can configure API key authentication on a Gateway by deploying an AuthPolicy and creating per-plan API key secrets, so that {prodname} can identify which rate-limiting plan applies to each request.

Reviewer: ```suggestion
You can configure API key authentication on a `Gateway` object by deploying an `AuthPolicy` custom resource (CR) and creating per-plan API key secrets, so that {prodname} can identify which rate-limiting plan applies to each request.
```

.Prerequisites

* You installed {prodname}.
* You installed {oc-first}.

> * You are logged in to your cluster.

Reviewer: do we need to be a cluster admin?

Author: Egutski confirmed that we do not.

* You created the `Gateway` object for your deployment.

.Procedure

> . Deploy an AuthPolicy that targets the Gateway and enables API key authentication by using the following example:

Reviewer: ```suggestion
. Create an `AuthPolicy` CR that targets the `Gateway` object and enables API key authentication by using the following example:
```

+
.Example API key AuthPolicy CR
[source,yaml,subs="+quotes"]
----
apiVersion: kuadrant.io/v1
kind: AuthPolicy
metadata:
  name: _<sample_app>_
  namespace: ${KUADRANT_GATEWAY_NS}
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: Gateway
    name: ${KUADRANT_GATEWAY_NAME}
  rules:
    authentication:
      "api-key-plan":
        apiKey:
          selector:
            matchLabels:
              app: _<sample_app.yaml>_
          allNamespaces: true
        credentials:
          authorizationHeader:
            prefix: APIKEY
----

> * Replace `_<sample_app>_` with the metadata for your application.

Reviewer: ```suggestion
+
* Replace `_<sample_app>_` with the metadata for your application.
```

> * Replace `_<sample_app.yaml>_` with the name of your application.

Reviewer: ```suggestion
```

* Replace 'KUADRANT_GATEWAY_NS' with the namespace for your gateway in {prodname}.
* Replace `KUADRANT_GATEWAY_NAME` with the name of your gateway in {prodname}.

> +

Reviewer: ```suggestion
```

* The `spec.rules.authentication."api-key-plan".apiKey.selector.matchLabels` field value points to the API key `Secrets` objects that carry the `app: _<sample_app>_` label.
* Setting `spec.rules.rules.authentication."api-key-plan".apiKey.selector.allNamespaces: true` allows secrets from any namespace to be matched.
+
. Apply the policy by running the following command:
+
[source,terminal,subs="+quotes"]
----

> $ oc apply -f

Reviewer 2: This command is incomplete, requires filename that contains the configuration to apply.

> $ oc apply -f

Reviewer: ```suggestion
$ oc apply -f _<sample_app.yaml>_
```

> ----

Reviewer: ```suggestion
----
+
* Replace `_<sample_app.yaml>_` with the name of your application.
```
